rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,124 @@
+[drawbacks]: #drawbacks
+
+Some possible drawbacks:
+
+- A new build system would require changes on the code
+  + On the other hand, such changes are likely to improve the code base.
+  
+- A new build system requires the developers become familiarized with it
+  - Specially when this build system uses its own description language.
+  + However, Meson is well documented, and its language is easy to grasp, specially for those familiarized with Python.
+
+- A new build system indirectly brings its own dependencies to the Nix project
+  - In particular, the reference implementation of Meson is written in Python.
+  - Further, this reference implementation generates script files meant to be consumed by Ninja, a tool written in C++ that acts like a Make replacement.
+  - This particular setting brings concerns about complexifying the bootstrap route.
+  + Given that Nix is currently written in C++, we can assume a C++ compiler as part of such a bootstrap route.
+  + There are full-featured alternative tools that replace Meson and Ninja. 
+    + Namely, Muon and Samurai are implementations of Meson and Ninja that require only a C compiler and a set of POSIX standard tools.
+
+- A new build system would require new strategies from the end users
+  - In particular, package managers that deploy Nix for their respective platforms.
+  + However, Meson and Ninja are nowadays a widespread toolset.
+    + Many open source projects use it, from mpv and dosbox-staging to Xorg and GNOME
+    + According to Repology, Meson is present in 53 package manager's families
+
+- The transition between between the old and new build systems should be smooth
+  - A wrapper script, maybe?
+  + Meson is not an obscure project; a mere documentation update should suffice
+
+# Alternatives
+[alternatives]: #alternatives
+
+The alternatives are
+
+- Doing nothing
+
+  It would keep the current code confusing and harder to work with.
+  
+- Other building systems (cmake, waf, premake etc.)
+  - Their strenghts and weaknesses should be evaluated.
+    - Tools strongly tied to other programming languages are strongly discouraged, because they further complexifies the bootstrap route as discussed above.
+      - Namely, waf is basically a Python library, whereas premake is a Lua library.
+    - Cmake has many noteworthy advantages:
+      + Can generates Make- and Ninja-compatible scripts;
+      + Supports Windows NT;
+      + Supports typical high level idiomatic constructions;
+      - On the other hand, the language is arguably more complex.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Questions that deserve further inquiry:
+
+- Unexpected interactions with Meson and Ninja
+  - Specially, vendoring and reproducibility.
+
+# Future work
+[future]: #future-work
+
+- Deprecate the current build scripts
+- Backport the new build system to Nix 2.3
+  - It was the latest release without Flakes support; it is important to bring such a deep modification to it.
+
```

Comment:
please add:

"replace bash-style buildsystem shell scripts with more portable alternatives"

Reply:
Why?

Reply:
not all distributions ship with bash, so a good possible goal would be to remove bash as a dependency and only rely on posix shell commands

Reply:
It wouldn't be solved by replacing everything with meson, after all?
Or there is something I am not catching?

Reply:
ill have to double check ..


there are some nix specific configuration bash scripts that are executed inside the buildsystem. moving these inside meson would mean that you would have to repeat those same steps to place the scripting inside of the makefiles/cmake/whatever too; or just have the sh file there still. imo there's no real benefit to embedding the shell commands in such a way, and keeping them separate makes them easier for users/devs to update/modify (they dont have to make the same change multiple times)

Reply:
Hum, I think I understood.

I believe this is implicitly included on the "better code structure", and will be a consequence of converting the build system.

Reply:
fair enough :)

Reply:
Also, related is the [RFC 99](https://github.com/NixOS/rfcs/pull/99) that proposes to implement oil as default stdenv shell. Among the issues are removing the bashisms around Nixpkgs.